projects/ADuCM3029_demo_cn0401/src/can_obj_layer.c

```diff
@@ -0,0 +1,1353 @@
+#include <stdlib.h>
+#include "can_obj_layer.h"
+#include "error.h"
+#include "delay.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * Get default configuration for the CAN controller initialization structure.
+ *
+ * @param [out] init - Pointer to the initialization structure.
+ *
+ * @return void.
+ */
+void can_ctrl_get_config(struct can_ctrl_init_param *init)
+{
+	init->can_ctrl_spi_init.chip_select = 0xFF;
+	init->can_ctrl_spi_init.extra = NULL;
+	init->can_ctrl_spi_init.id = SPI_ARDUINO;
+	init->can_ctrl_spi_init.max_speed_hz = 4000000;
+	init->can_ctrl_spi_init.mode = SPI_MODE_0;
+	init->can_ctrl_spi_init.type = ADICUP3029_SPI;
+	init->can_dbt = BITRATE_DBT_2M;
+	init->can_nbt = BITRATE_NBT_500K;
+	init->con_iso_crc_en = true;
+	init->con_store_in_tef_en = true;
+	init->con_txq_en = true;
+	init->rx_fifo_nr = 2;
+	init->rx_fifo_plsize = DATA_BYTES_64;
+	init->rx_fifo_size = 1;
+	init->rx_fifo_tsen = true;
+	init->rx_flt_nr = 0;
+	init->rx_sid_addr = 0x300;
+	init->ssp_mode = SSP_MODE_AUTO;
+	init->tef_fifo_size = 10;
+	init->tef_time_stamp_en = true;
+	init->tx_fifo_nr = 1;
+	init->tx_fifo_plsize = DATA_BYTES_64;
+	init->tx_fifo_priority = 1;
+	init->tx_fifo_size = 1;
+	init->txq_fifo_size = 2;
+	init->txq_plsize = DATA_BYTES_64;
+	init->txq_tx_priority = 0;
+}
+
+/**
+ * Read a byte from the CAN controller memory using the SPI interface.
+ *
+ * Only Special Function Registers (SFRs) can be read at byte level.
+ *
+ * @param [in]  dev     - Pointer to the handler structure.
+ * @param [in]  address - The address of the byte to be read.
+ * @param [out] data    - Pointer to the data container.
+ *
+ * @return 0 in case of success, error code in case of failure.
+ */
+int32_t can_ctrl_sfr_byte_read(struct can_ctrl_dev *dev, uint16_t address,
+			       uint8_t *data)
+{
+	uint8_t spi_msg[] = {0, 0, 0};
+	int32_t ret;
+
+	spi_msg[0] |= CAN_CTRL_CMD_MODE(CAN_CTRL_CMD_READ);
+	spi_msg[0] |= CAN_CTRL_ADDR_UP_NIBBLE_MODE(address);
+	spi_msg[1] |= CAN_CTRL_ADDR_DW_BYTE_MODE(address);
+
+	ret = spi_write_and_read(dev->can_ctrl_spi, spi_msg, 3);
+	if(ret != SUCCESS)
+		return ret;
+
+	*data = spi_msg[2];
+
+	return ret;
+}
+
+/**
+ * Write a byte in the CAN controller memory using the SPI interface.
+ *
+ * Only Special Function Registers (SFRs) can be written at byte level.
+ *
+ * @param [in] dev     - Pointer to the handler structure.
+ * @param [in] address - The address of the byte to be written.
+ * @param [in] data    - New value of the byte.
+ *
+ * @return 0 in case of success, error code in case of failure.
+ */
+int32_t can_ctrl_sfr_byte_write(struct can_ctrl_dev *dev, uint16_t address,
+				uint8_t data)
+{
+	uint8_t spi_msg[] = {0, 0, 0};
+
+	spi_msg[0] |= CAN_CTRL_CMD_MODE(CAN_CTRL_CMD_WRITE);
```

Comment:
these two can be done in one go:
```
spi_msg[0] = CAN_CTRL_CMD_MODE(CAN_CTRL_CMD_WRITE) |
CAN_CTRL_ADDR_UP_NIBBLE_MODE(address);
```
Also for the others you can remove the OR operator.